Resources/ProxyProvider/DIRACCAProxyProvider.py

```diff
@@ -33,227 +66,249 @@ def __init__(self, parameters=None):
+    if kwargs.get('Email'):
+      kwargs['emailAddress'] = [kwargs['Email']]
 
-    # Fill DN subject name
-    if not userDict.get('FullName'):
-      return S_ERROR("Incomplete user information: no full name found")
-    if not userDict.get('Email'):
-      return S_ERROR("Incomplete user information: no email found")
     self.__X509Name = X509.X509_Name()
-    self.log.info('Creating distributes names chain')
-    # Test match fields
-    for nid in self.match:
-      if nid not in caFieldByNid:
-        return S_ERROR('Distributes name(%s) must be present in CA certificate.' % ', '.join(self.n2fields[nid]))
-      result = self.__fillX509Name(caFieldByNid[nid], caDict[caFieldByNid[nid]])
+    self.log.info('Creating distinguished names chain')
+
+    for nid in self.supplied:
+      if nid not in [self.fields2nid[f] for f in self.dnList]:
+        return S_ERROR('DNs order list does not contain supplied DN "%s"' % self.nid2field[nid])
+
+    for field in self.dnList:
+      values = []
+      nid = self.fields2nid[field]
+      if nid in self.match:
+        for field in self.nid2fields[nid]:
+          if field in self.dnInfoDictCA:
+            values = self.dnInfoDictCA[field]
+        if not values:
+          return S_ERROR('Not found "%s" match DN in CA' % field)
+      for field in self.nid2fields[nid]:
+        if kwargs.get(field):
+          values = kwargs[field] if isinstance(kwargs[field], list) else [kwargs[field]]
+      if not values:
+        for field in self.nid2fields[nid]:
+          if self.parameters.get(field):
+            values = self.parameters[field]
+      if not values and nid in self.supplied:
+        return S_ERROR('No values set for "%s" DN' % field)
+
+      result = self.__fillX509Name(field, values)
       if not result['OK']:
         return result
-    # Test supplied fields
-    for nid in self.supplied:
-      if self.defDict.get(self.n2field[nid]):
-        result = self.__fillX509Name(self.n2field[nid], self.defDict[self.n2field[nid]])
-        if not result['OK']:
-          return result
-    for nid, value in [(self.fs2nid['CN'], userDict['FullName']),
-                       (self.fs2nid['emailAddress'], userDict['Email'])]:
-      if nid in self.supplied + self.optional:
-        result = self.__fillX509Name(self.n2field[nid], value)
-        if not result['OK']:
-          return result
 
     # WARN: This logic not support list of distribtes name elements
     resDN = m2.x509_name_oneline(self.__X509Name.x509_name)  # pylint: disable=no-member
-    if userDN and not userDN == resDN:
-      return S_ERROR('%s not match with generated DN: %s' % (userDN, resDN))
+
+    result = self.checkStatus(resDN)
+    if not result['OK']:
+      return result
     return S_OK(resDN)
 
   def __parseCACFG(self):
     """ Parse CA configuration file
     """
     block = ''
     self.cfg = {}
+    self.supplied, self.optional, self.match, self.dnList = [], [], [], []
```

Comment:
`self.supplied = self.optional = self.match = self.dnList = []`

Reply:
I really do not understand why need to do that?

Reply:
in fact what I proposed is even wrong, because it is a mutable type, so that would not do what you want !